[MASWE-0020] Weak Encryption (by @appknox) (#2910)
OWASP/owasp-mastg: weaknesses/MASVS-CRYPTO/MASWE-0020.md

For example the use of TDEA (Triple Data Encryption Algorithm), which is often referred to as "Triple DES" or "3DES" is disallowed by NIST since end of 2023, due to known attacks like, "meet-in-the-middle", collision attacks and [Sweet32](https://nvd.nist.gov/vuln/detail/CVE-2016-2183).

Similarly, hash functions such as MD5 and SHA-1, which were previously popular for ensuring data integrity, are now considered deprecated as attackers can generate hash collisions, potentially leading to data tampering.

## Impact

- **Loss of Confidentiality**: Weak encryption may enable attackers to decipher and obtain sensitive information, resulting in unauthorized exposure and possible data breaches.

- **Loss of Integrity**: Weak encryption can compromise the integrity of data, allowing adversaries to alter or manipulate the information without detection.

## Mode of Introduction

> - **Use of Deprecated Algorithms** : Relying on outdated or weak cryptographic algorithms can allow threat actors to attack the cipher text, key or exploit known vulnerabilities in the algorithm, for example through brute force attacks.

Reviewer: ```suggestion
- **Use of Deprecated Algorithms** : Relying on outdated or weak encryption algorithms can allow threat actors to attack the cipher text, key or exploit known vulnerabilities in the algorithm, for example through brute force attacks.
```

---

## Overview

> Weak encryption denotes cryptographic systems or implementations that exhibit vulnerabilities to attacks, allowing unauthorized individuals to decrypt secured data with minimal effort. This inadequacy may arise from various reasons, such as, use of outdated algorithms, short key lengths, weak encryption modes(ECB) and improper implementation practices. For instance, the DES encryption standard, once considered robust, is now vulnerable to brute-force attacks because its key sizes are too small to withstand the capabilities of contemporary computing power. In a similar, hash functions like MD5 and SHA-1, which were previously popular for ensuring data integrity, are now deemed weak as attackers can easily generate hash collisions, resulting in security breaches.

Reviewer 2: Let's focus on AES and ECB that shouldn't be used, which is also highlighted in the android and other documentations instead of DES.

```suggestion
Weak encryption refers to cryptographic systems or implementations that are vulnerable to attack, allowing unauthorised individuals to decrypt secured data. This weakness can be due to a number of reasons, including the use of outdated algorithms, deprecated encryption modes such as ECB and improper implementation practices such as the use of a non-random or empty Initialisation Vector (IV). 

For example the use of TDEA (Triple Data Encryption Algorithm), which is often referred to as "Triple DES" or "3DES" is disallowed by NIST since end of 2023, due to known attacks like, "meet-in-the-middle", collision attacks and [Sweet32](https://nvd.nist.gov/vuln/detail/CVE-2016-2183).

Similarly, hash functions such as MD5 and SHA-1, which were previously popular for ensuring data integrity, are now considered deprecated as attackers can generate hash collisions, potentially leading to data tampering.
```

## Impact

> - **Loss of  Confidentiality**: Weak encryption may enable attackers to decipher and obtain sensitive information, resulting in unauthorized exposure and possible data breaches.

Reviewer 2: ```suggestion
- **Loss of Confidentiality**: Weak encryption may enable attackers to decipher and obtain sensitive information, resulting in unauthorized exposure and possible data breaches.
```

> - **Loss of Integrity**: Weak encryption may compromise the integrity of data, allowing adversaries to exploit vulnerabilities and potentially alter or manipulate the information without detection.

Reviewer 2: ```suggestion
- **Loss of Integrity**: Weak encryption can compromise the integrity of data, allowing adversaries to alter or manipulate the information without detection.
```

## Mode of Introduction

> - **Use of Deprecated Algorithms** : Relying on outdated or weak cryptographic algorithms can produce keys that are less secure. Such algorithms are often characterized by vulnerabilities or the ability to use shorter key lengths, which increases their susceptibility to contemporary attacks and threatens the overall security of the application.

Reviewer 2: Let's make this more granular: 

```suggestion
- **Use of Deprecated Algorithms** : Relying on outdated or weak cryptographic algorithms can allow threat actors to attack the cipher text, key or exploit known vulnerabilities in the algorithm, for example through brute force attacks.
- **Insecure Modes of Operation**:  Using modes that are considered deprecated increase the attack surface of encrypted information. For example the use of AES/ECB is deprecated as it divides the plaintext into blocks and encrypts each block separately using the same key. This makes the cipher text vulnerable to "known plaintext attacks" and leaks information about the structure of the original plaintext. 
- **Predictable Initialization Vectors (IVs)**: If IVs are not random or unique, they can be exploited in attacks like ciphertext injection or pattern recognition. This compromises the confidentiality of encrypted data, especially in modes like CBC (Cipher Block Chaining).
- **Weak Keys**: Short or easily guessable keys compromise encryption strength. The use of small key sizes (e.g., 56-bit keys in DES) can make the encryption susceptible to brute-force attacks. Best practices recommend keys of at least 256 bits for strong encryption.
- **Misuse of Non-Cryptographic Operations**: Relying on techniques such as XOR, Base64 encoding, or simple obfuscation methods for security purposes. These methods provide no actual encryption and can be easily reversed or decoded, exposing sensitive data.
```

## Mitigations

> - Always use modern, well-established cryptographic libraries and APIs that follow best practices for entropy generation and key management.

Reviewer 2: ```suggestion
- Always use modern, well-established cryptographic libraries in mobile apps that follow best practices and offer cryptographic algorithms that are aligned with the recommendations by NIST.
```

> - It is recommended to use secure modes of Operation like CBC (Cipher Block Chaining).

Reviewer 2: ```suggestion
- For example it is recommended to use secure modes of operations like GCM (Galois/Counter Mode) mode for AES with 256-bit keys, providing both encryption and authentication.
```
